### PR TITLE
[nginx-ingress-controller] Add support for default backend in Ingress rule

### PR DESCRIPTION
**Pull request by aledbf**
replaces #1759

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1830)

### CODE REVIEW

**Comment by aledbf** [comment]
@bprashanth ready for a first review

File: ingress/controllers/nginx/controller.go [hunk 1]
```diff
@@ -739,19 +740,38 @@ func (lbc *loadBalancerController) getUpstreamServers(ngxCfg config.Configuratio
 				server = servers[defServerName]
 			}
 
+			if rule.HTTP == nil && host != defServerName {
+				glog.Info("multiple root ISSUE")
```
Comment (bprashanth) [inline ingress/controllers/nginx/controller.go]: maybe send an event, and turn the message into a WARNING or ERROR? Please either extend the message itself or add a comment.

File: ingress/controllers/nginx/controller.go [hunk 2]
```diff
@@ -739,19 +740,38 @@ func (lbc *loadBalancerController) getUpstreamServers(ngxCfg config.Configuratio
...
+						ups = defUps
+					}
+				}
+
 				nginxPath := path.Path
 				// if there's no path defined we assume /
 				if nginxPath == "" {
 					lbc.recorder.Eventf(ing, api.EventTypeWarning, "MAPPING",
-						"Ingress rule '%v/%v' contains no path definition. Assuming /", ing.GetNamespace(), ing.GetName())
+						"Ingress rule '%v/%v' contains no path definition. Assuming /",
```
Comment (bprashanth) [inline ingress/controllers/nginx/controller.go]: / is the same as /* right? do you lose anything by just making it the latter (more explicit)

File: ingress/controllers/nginx/controller.go [hunk 3]
```diff
@@ -739,19 +740,38 @@ func (lbc *loadBalancerController) getUpstreamServers(ngxCfg config.Configuratio
...
+						ing.GetNamespace(), ing.GetName())
 					nginxPath = rootLocation
 				}
 
-				// Validate that there is no another previuous
+				// Validate that there is no another previous
```
Comment (bprashanth) [inline ingress/controllers/nginx/controller.go]: s/no another/no/

File: ingress/controllers/nginx/controller.go [hunk 4]
```diff
@@ -844,41 +880,54 @@ func (lbc *loadBalancerController) createUpstreams(ngxCfg config.Configuration,
...
-						break
-					}
-				}
+	var upstreams []ingress.UpstreamServer
+	if err != nil {
+		return upstreams, fmt.Errorf("error getting service %v from the cache: %v", svcKey, err)
+	}
+
+	if !svcExists {
+		return upstreams, fmt.Errorf("service %v does not exists", svcKey)
```
Comment (bprashanth) [inline ingress/controllers/nginx/controller.go]: do these translate into events? (the gce controller has the same problem of not sending an event for this case, pretty confusing to debug)

File: ingress/controllers/nginx/nginx/command.go [hunk 5]
```diff
@@ -128,7 +128,10 @@ func (ngx Manager) testTemplate(cfg []byte) error {
 		return err
 	}
 	defer tmpfile.Close()
-	defer os.Remove(tmpfile.Name())
+	// do not remove temporal nginx.conf file if log level is >= 3
+	if !glog.V(3) {
```
Comment (bprashanth) [inline ingress/controllers/nginx/nginx/command.go]: hmm, why? this looks like a weird lateral dependency. are we just leaking the file to debug later?

File: ingress/controllers/nginx/controller.go (hunk 1 above)
Comment (bprashanth) [inline ingress/controllers/nginx/controller.go]: This is actually worth documenting. We just enforce that if there're multiple ingresses asking for a default backend, they must have the same backend? do we pick this backed on fcfs basis?

Comment (aledbf) [inline ingress/controllers/nginx/controller.go]: done.

File: ingress/controllers/nginx/controller.go (hunk 2 above)
Comment (aledbf) [inline ingress/controllers/nginx/controller.go]: done, comment added

File: ingress/controllers/nginx/controller.go (hunk 3 above)
Comment (aledbf) [inline ingress/controllers/nginx/controller.go]: done

File: ingress/controllers/nginx/controller.go (hunk 4 above)
Comment (aledbf) [inline ingress/controllers/nginx/controller.go]: no but this appears in the log of the pod

File: ingress/controllers/nginx/nginx/command.go (hunk 5 above)
Comment (aledbf) [inline ingress/controllers/nginx/nginx/command.go]: change the behavior to remove the file only if there's no error running the check

**Comment by bprashanth** [comment]
LGTM sorry for the delay

**Pull request merged**